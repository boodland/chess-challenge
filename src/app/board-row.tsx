import { useContext } from 'react';
import BoardCellMemoized from './board-cell';
import { COLUMNS, CellTypeEnum, isEmpty } from './utils';
import { GameContext } from './game-context';

const getCellColors = (row: number) => {
  return (row % 2 !== 0) ?
      [CellTypeEnum.light, CellTypeEnum.dark] :
      [CellTypeEnum.dark, CellTypeEnum.light];
}

export interface BoardRowProps {
  startPosition: number;
}

function BoardRow({ startPosition }: BoardRowProps) {
  const { positions, selectedPosition, onClickPosition } =
    useContext(GameContext);
  const boardRowPosition = COLUMNS.map(
    (column, index) => index + startPosition * COLUMNS.length
  );
  const cellColors = getCellColors(startPosition);
  const boardRowCells = boardRowPosition.map((position) => {
    const piece = positions[position];
    const isBlocked = isEmpty(piece) && typeof selectedPosition === 'undefined';
    const onClickHandler = isBlocked
      ? () => undefined
      : () => onClickPosition(position);
    return (
      <BoardCellMemoized
        key={position}
        piece={piece}
        type={cellColors[position % 2]}
        blocked={isBlocked}
        selected={selectedPosition === position}
        onClick={onClickHandler}
      />
    );
  });
  return <div>{boardRowCells}</div>;
}

export default BoardRow;
