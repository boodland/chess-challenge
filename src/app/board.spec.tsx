import { render } from '@testing-library/react';

import Board from './board';
import { BoardRowProps } from './board-row';

const { ROWS, COLUMNS } = vi.hoisted(() => {
  return { ROWS: [0, 1, 2, 3, 4], COLUMNS: ['a', 'b', 'c'] }
})

vi.mock('./utils', () => {
  return { ROWS, COLUMNS };
});

const mockedComponentName = 'MockedBoardRow';

vi.mock('./board-row', () => ({
  default: (props: BoardRowProps) => (
    <div>{`${mockedComponentName} ${JSON.stringify(props)}`}</div>
  ),
}));

describe('Board', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <Board/>
    );
    expect(baseElement).toBeTruthy();
  });

  it('should have the correct styles', () => {
    const { container } = render(
      <Board/>
    );
    expect(container).toMatchSnapshot();
  });

  it('should display columns row', () => {
    const { getAllByText } = render(
      <Board/>
    );
    expect(getAllByText(new RegExp('^[abc]$', "i"))).toHaveLength(COLUMNS.length);
  });

  it('should display one BoardRow per row', () => {
    const { getAllByText } = render(
      <Board/>
    );
    expect(getAllByText(mockedComponentName, {exact: false})).toHaveLength(ROWS.length);
  });
});
