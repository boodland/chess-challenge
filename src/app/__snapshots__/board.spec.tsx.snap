// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Board > should have the correct styles 1`] = `
<div>
  <div>
    <span
      class="_text_672456"
    >
      5
    </span>
    <div>
      MockedBoardRow {"startPosition":0}
    </div>
  </div>
  <div>
    <span
      class="_text_672456"
    >
      4
    </span>
    <div>
      MockedBoardRow {"startPosition":1}
    </div>
  </div>
  <div>
    <span
      class="_text_672456"
    >
      3
    </span>
    <div>
      MockedBoardRow {"startPosition":2}
    </div>
  </div>
  <div>
    <span
      class="_text_672456"
    >
      2
    </span>
    <div>
      MockedBoardRow {"startPosition":3}
    </div>
  </div>
  <div>
    <span
      class="_text_672456"
    >
      1
    </span>
    <div>
      MockedBoardRow {"startPosition":4}
    </div>
  </div>
  <div
    class="_columns_672456"
  >
    <span
      class="_text_672456"
    >
      a
    </span>
    <span
      class="_text_672456"
    >
      b
    </span>
    <span
      class="_text_672456"
    >
      c
    </span>
  </div>
</div>
`;
